concent_api/utils/helpers.py

```diff
@@ -46,3 +48,46 @@ def parse_datetime_to_timestamp(date_time: datetime.datetime) -> int:
         return int(date_time.replace(tzinfo = timezone.utc).timestamp())
     else:
         return int(date_time.astimezone(timezone.utc).timestamp())
+
+
+def get_task_id_from_message(golem_message: message.base.Message) -> str:
+    """ Returns task_id nested inside given Golem Message depending on its type. """
+
+    if isinstance(golem_message,            message.ComputeTaskDef):
+        return golem_message['task_id']
+
+    elif isinstance(golem_message,          message.TaskToCompute):
+        return golem_message.compute_task_def['task_id']
+
+    elif isinstance(golem_message,          message.concents.ForceGetTaskResult):
+        return golem_message.report_computed_task.task_to_compute.compute_task_def['task_id'],
+
+    elif isinstance(golem_message,          message.concents.ForceGetTaskResultUpload):
+        return golem_message.force_get_task_result.report_computed_task.task_to_compute.compute_task_def['task_id']
+
+    elif isinstance(golem_message, (
+                                            message.VerdictReportComputedTask,
+                                            message.concents.ForceSubtaskResults,
+    )):
+        return golem_message.ack_report_computed_task.task_to_compute.compute_task_def['task_id']
+
+    elif isinstance(golem_message, (
+                                            message.ForceReportComputedTask,
+                                            message.AckReportComputedTask,
+                                            message.concents.ForceGetTaskResultFailed,
+    )):
+        return golem_message.task_to_compute.compute_task_def['task_id']
+
+    elif isinstance(golem_message,          message.RejectReportComputedTask):
+        if isinstance(golem_message.cannot_compute_task, message.tasks.CannotComputeTask):
+            return golem_message.cannot_compute_task.task_to_compute.compute_task_def['task_id']
+        elif isinstance(golem_message.task_to_compute, message.tasks.TaskToCompute):
+            return golem_message.task_to_compute.compute_task_def['task_id']
+
+    elif isinstance(golem_message,          message.concents.ForceSubtaskResultsResponse):
+        if isinstance(golem_message.subtask_results_accepted,   message.tasks.SubtaskResultsAccepted):
+            return golem_message.subtask_results_accepted.subtask_id
+        elif isinstance(golem_message.subtask_results_rejected, message.tasks.SubtaskResultsRejected):
+            return golem_message.subtask_results_rejected.report_computed_task.subtask_id
+
+    return None
```

Comment:
Please note that currently we allow to send any message to `send`, and we want to log it, so actually this function needs to cover every other golem_message, and this list need to be updated on each added message, which is not really good way to maintain it.

Reply:
But you already need to look at this function every time someone updates golem-messages because there might be a new message class that contains the task ID. Without a check like this no one will remember to do this.

Reply:
But you are aware how much work it is now and on each change of golem_mesages ? There can be message which have a nested message which have a nested message which have a message containing `task_id`. On each change you have to analyze all messages and their connections and update this function. 

Because if I only add all other messages than those currently used in this function to big `if` containing a list of all other messages and `return None`, and we will start using one of those messages, then it will be as you said, no one will notice that it is not logging correctly before looking to logs.

Also, if I just want to add `assert False` and `raise` at the end, I can't because it means that Concent stops supporting messages without task_id (because it will fail on logging).

What I can propose is to remove debug logging and log whole message on each case when `task_id` is needed.

Reply:
1) Look, your current code looks like this:

    ```
    if A: return A.id
    if B: return B.C.id
    otherwise: return None
    ```

    The problem with a blacklist like this is that whenever something new is added, your code won't handle it properly and no one will notice. Say, we add message `D` with the ID field. Your code ignores it and still returns `None`.

    So how do you ensure that your code is not broken after golem-messages adds message `D`? You can't. On every update of golem messages you have go through all messages in golem-messages and see if there are any new ones and if any of the old ones have changed.

    **You already have to do all this work because you used a blacklist**.

    What I propose is to use a whitelist. You list everything you know and if anything from outside of that list comes up, you notify the programmer that he has to update the code:

    ```
    if A: return A.id
    if B: return B.C.id
    otherwise: assert False
    ```

    Now if golem-messages adds `D`, you get an error and know that you have to fix your code. This should be the default approach.

    ```
    if A: return A.id
    if B: return B.C.id
    if D: return None
    otherwise: fail
    ```

    This is actually less work for you in the long term (even if you have to list all the messages once) because Python can detect and fail on newly added messages. You don't have to review the source of golem-messages yourself to find them.

    It still does not solve the problem of someone changing `D` to add the ID later. But a blacklist does not help with that case either so it's still better than doing nothing.

2) I now see that there's something I did not think about: what about custom messages with arbitrary `TYPE`. After all a type is just an int and someone could create a message with a type that's not listed in golem-messages. The code should not fail in that case.

    This is actually a good argument against the whitelist approach. But this only means that you'll have to maintain this function manually. I think this code will often silently break and I don't like that. We have to find a sensible solution here that won't rely on someone regularly going over golem-messages source and finding changes in the ID fields.

> What I can propose is to remove debug logging and log whole message on each case when `task_id` is needed.

I don't like this because messages could be large. Most of the time we don't need the whole message, just the important parts. This has potential to make the logs useless because of too much information.

I have a different solution: instead of hard-coding the locations of ID fields, recursively go over all the nested messages and in each one look for a field called `task_id` and `subtask_id`. If it's there, use it.

Reply:
:+1: